### ISSUE TITLE
Trigger support for Apprise

### ISSUE THREAD

**Issue by w0ndersp00n** [fmartinou/whats-up-docker#60]
Would it be possible to add support for triggering Apprise (https://github.com/caronc/apprise-api / https://github.com/caronc/apprise)? It basically is a gateway for various notification services, including some that are already supported.

The big difference is that Apprise requires a body, which then is sent to the configured notification services.

<img width="861" alt="Schermafbeelding 2021-08-08 om 23 13 31" src="https://user-images.githubusercontent.com/15470927/128645876-98f992f0-a5cf-45c1-9076-85bbc6743900.png">

This would enable a lot of new notification possibilities, but it might decrease the level of flexibility.

**Comment by fmartinou** [comment]
Hi,

I took a look at Apprise at the beginning of WUD because the list of the available integrations with Apprise is just insane 🤩 !
And unfortunately, Apprise is a Python lib whereas WUD is written in Javascript.

Now I can see that Apprise creator also provides a standalone Docker image that allows Apprise to be used using an HTTP API.

The potential is appealing to me but I I don't really know how to integrate 🤔 .

I now see 2 solutions:

### 1. Embed Apprise into WUD image
It'd mean:
- Install Python in WUD image
- Install Apprise in WUD image
- Make WUD run [Apprise CLI](https://github.com/caronc/apprise/wiki/CLI_Usage) to notify

Cons:
- Bigger WUD image with extra dependencies (Python...)

### 2. Delegate to [Apprise API](https://github.com/caronc/apprise-api)
- Make WUD call Apprise API to notify

Cons:
- Not straightforward for users (because it means running an extra container...)

What do you think about that?
Which solution have you in mind when raising the issue?
Solution 1, Solution 2 or maybe an other Solution I forgot?

**Comment by w0ndersp00n** [comment]
I think I'd prefer option 2. This way WUD doesn't grow bigger then it needs to I think. It also enables users to choose which solution they use. If you're using Apprise already, then it should be possible to use the existing container to handle all logic. If you don't want to use Apprise, then you still have the WUD integrated solutions.

I also saw how Bazarr integrated Apprise within their app (option 1). I don't think that similar integration would be great for WUD:

- It might require a UI to configure them (so do what Apprise already is providing)
- Or, you'd have to duplicate your notification URL using labels in order to send out the correct notification

In cases where multiple containers are used and are trying to send notifications, e.g. via SMTP or Slack, it would be easier to manage the settings in the gateway, and let the container call the Apprise container.

I think the optimal approach would be to have the body/title/tag/URL as labels, which keeps configuration in Apprise, but lets WUD call it.

So something like this:

`- WUD_TRIGGER_APPRISE_URL=http://apprise:8004
- WUD_TRIGGER_APPRISE_BODY=...
- WUD_TRIGGER_APPRISE_TITLE=...
- WUD_TRIGGER_APPRISE_TAG=...`

I don't know if that makes sense and would be possible?

**Comment by fmartinou** [comment]
> I don't know if that makes sense and would be possible?

Sure! (I do agree with your answer).

With that solution, WUD will benefit from tons of integrations for (almost) free (I just need to find some time to implement it ;)

I'll keep you updated.

**Comment by fmartinou** [comment]
I have a first version: `fmartinou/whats-up-docker:feature__60`

For now, I use the Apprise API in the [stateless way](https://github.com/caronc/apprise-api#stateless-solution) so I don't support tags but IMHO it's easier to integrate for users because there is no need to configure anything on Apprise.

WUD configuration is really simple ([see the draft of the documentation here](https://github.com/fmartinou/whats-up-docker/tree/feature/%2360/docs/configuration/triggers/apprise)):
- 1 env var for the Apprise API url
- 1 env var for the Apprise notification urls

Example:
```yaml
version: '3'

services:
  whatsupdocker:
    image: fmartinou/whats-up-docker
    ...
    environment:
      - WUD_TRIGGER_APPRISE_LOCAL_URL=http://apprise:8000
      - WUD_TRIGGER_APPRISE_LOCAL_URLS=mailto://john.doe:secret@gmail.com,sns://AHIAJGNT76XIMXDBIJYA/bu1dHSdO22pfaaVy/wmNsdljF4C07D3bndi9PQJ9/us-east-2/+1(800)555-1223
```

I don't allow to provide a template for the title & the body (as well as for the other wud integrations) but it could be an improvement.

Example of notification content:
```
# Title
[WUD] New tag found for container radarr

# Body
Container radarr running with tag 3.2.1.5070-ls105 can be updated to tag 3.2.2.5080-ls112
```

**Comment by w0ndersp00n** [comment]
This looks great! Let me pull the image and give it a spin.

I think the notification content is is enough already; I mean it tells that an update is available.

**Comment by w0ndersp00n** [comment]
It seems to be working great! I've received four notifications for updates containers on the configured endpoints :)

**Comment by fmartinou** [comment]
Great!

I'm gonna merge it and release a new `5.3.0` version.

**Comment by fmartinou** [comment]
Available in `5.3.0`.
